File: packages/aegis-sdk-py/salanor_aegis_ledger/record.py
from __future__ import annotations

import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from salanor_aegis_ledger.canonical import event_hash_from_body
from salanor_aegis_ledger.schema import validate_event
from salanor_aegis_ledger.store import append_event, last_event_hash, read_events

DEFAULT_SIGNATURE = {
    "alg": "local-placeholder",
    "value": "placeholder:0000000000000000000000000000000000000000000000000000000000000000",
}


def record(
    store_path: Path | str,
    payload: dict[str, Any],
    *,
    recorded_at: str | None = None,
    event_id: str | None = None,
) -> dict[str, Any]:
    path = Path(store_path)
    existing = read_events(path)
    prev = last_event_hash(existing)

    body: dict[str, Any] = {
        "aps_version": "0.1",
        "event_id": event_id or str(uuid.uuid4()),
        "recorded_at": recorded_at or datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "tenant_id": payload.get("tenant_id", "local"),
        "actor": payload["actor"],
        "action": payload["action"],
        "subject": payload["subject"],
        "context": payload["context"],
        "signature": payload.get("signature", DEFAULT_SIGNATURE),
        "chain": {"prev_event_hash": prev},
    }

    event_hash = event_hash_from_body(body)
    event = {**body, "chain": {"prev_event_hash": prev, "event_hash": event_hash}}
    validate_event(event)
    append_event(path, event)
    return event

File: packages/aegis-ledger-sdk/src/record.ts
import { randomUUID } from "node:crypto";
import { eventHashFromBody } from "./canonical.js";
import { validateEvent } from "./schema.js";
import { appendEvent, lastEventHash, readEvents } from "./store.js";
import type { ApsEvent, ApsSignature, RecordInput, RecordOptions } from "./types.js";

const DEFAULT_SIGNATURE: ApsSignature = {
  alg: "local-placeholder",
  value: "placeholder:0000000000000000000000000000000000000000000000000000000000000000",
};

export interface RecordResult {
  event: ApsEvent;
  store_path: string;
}

/**
 * Append a validated APS-1 event to a local NDJSON ledger with hash chaining.
 */
export function record(
  storePath: string,
  input: RecordInput,
  options: RecordOptions = {},
): RecordResult {
  const existing = readEvents(storePath);
  const prev = lastEventHash(existing);

  const body: Omit<ApsEvent, "chain"> & { chain: { prev_event_hash: string | null } } = {
    aps_version: "0.1",
    event_id: options.event_id ?? randomUUID(),
    recorded_at: options.recorded_at ?? new Date().toISOString(),
    tenant_id: input.tenant_id ?? "local",
    actor: input.actor,
    action: input.action,
    subject: input.subject,
    context: input.context,
    signature: input.signature ?? DEFAULT_SIGNATURE,
    chain: { prev_event_hash: prev },
  };

  const event_hash = eventHashFromBody(body as unknown as Record<string, unknown>);
  const event: ApsEvent = {
    ...body,
    chain: { prev_event_hash: prev, event_hash },
  };

  validateEvent(event);
  appendEvent(storePath, event);
  return { event, store_path: storePath };
}
